src/directives/column-reorder.js

```diff
@@ -0,0 +1,223 @@
+(function(angular) {
+    "use strict";
+
+    var module = angular.module('table.column-reorder', ['net.enzey.services.events']);
+
+    module.directive('reorderableColumns', function() {
+        var incrementedId = 0;
+        var getId = function() {
+            return 'colId' + incrementedId++;
+        };
+
+        return {
+            controller: function($scope) {
+                var templateMapper = {};
+
+                this.getTemplate = function(rowId) {
+                    return templateMapper[rowId];
+                };
+
+                this.setTemplate = function(rowId, template) {
+                    if (!templateMapper[rowId]) {
+                        templateMapper[rowId] = template;
+                    }
+                };
+
+                this.reorderColumns = function(oldColIndex, newColIndex, offset) {
+
+                    if (oldColIndex === null || newColIndex === null) {
+                        return;
+                    }
+
+                    oldColIndex = +oldColIndex;
+                    newColIndex = +newColIndex;
+
+                    if (oldColIndex < 0 || newColIndex < 0) {
+                        return;
+                    }
+
+                    if (oldColIndex < newColIndex) {
+                        newColIndex--;
+                    }
+
+                    Object.keys(templateMapper).forEach(function(rowId) {
+                        var rowTemplate = templateMapper[rowId];
+                        rowTemplate = angular.element('<tr>' + rowTemplate + '<tr>');
```

Comment:
I didn't say it was perfect. The issue with the columns is I cannot use the same directive used on the headers and body elements unless there is always one and only one colgroup element. Since the columns just float around the styles would need to be ripped from a col element and placed on another, instead of moving the elements.